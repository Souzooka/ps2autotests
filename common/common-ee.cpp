#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
