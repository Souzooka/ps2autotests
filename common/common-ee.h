#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
