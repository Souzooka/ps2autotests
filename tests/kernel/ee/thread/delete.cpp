#include <common-ee.h>
#include <kernel.h>
#include <string.h>

#define STACK_SIZE 0x8000
#define TEST_THREAD_PRIORITY 0x40

char nullThreadStack[STACK_SIZE] __attribute__ ((aligned(16)));
void nullThreadProc(u32) {
	//Do nothing
}

int createTestThread(void *entry, int prio, void *stack, u32 stackSize) {
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func = entry;
	threadParam.initial_priority = prio;
	threadParam.stack = stack;
	threadParam.stack_size = stackSize;
	return CreateThread(&threadParam);
}

void doDeleteThread(int threadId) {
	int result = DeleteThread(threadId);
	if(result >= 0) {
		schedf("succeeded -> result: %s\n", (result == threadId) ? "thread id" : "something else");
	} else {
		schedf("failed -> result: %d\n", result);
	}
}

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");

	{
		schedf("self thread:\n");

		schedf("  delete (with current tid): ");
		doDeleteThread(GetThreadId());
		
		flushschedf();
	}

	{
		schedf("low prio thread:\n");
		
		int threadId = createTestThread((void*)&nullThreadProc, TEST_THREAD_PRIORITY + 0x10, nullThreadStack, STACK_SIZE);
		schedf("  delete before start: ");
		doDeleteThread(threadId);

		threadId = createTestThread((void*)&nullThreadProc, TEST_THREAD_PRIORITY + 0x10, nullThreadStack, STACK_SIZE);
		StartThread(threadId, NULL);
		schedf("  delete after start: ");
		doDeleteThread(threadId);

		flushschedf();
	}

	{
		schedf("high prio thread:\n");
		
		int threadId = createTestThread((void*)&nullThreadProc, TEST_THREAD_PRIORITY - 0x10, nullThreadStack, STACK_SIZE);
		StartThread(threadId, NULL);
		schedf("  delete after start: ");
		doDeleteThread(threadId);

		flushschedf();
	}

	//Various corner cases
	{
		schedf("corner cases:\n");
		
		schedf("  delete with invalid tid: ");
		doDeleteThread(~0);

		flushschedf();
	}
	
	printf("-- TEST END\n");
}
