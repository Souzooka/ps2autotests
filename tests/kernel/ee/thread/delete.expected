-- TEST BEGIN
self thread:
  delete (with current tid): failed -> result: -1
worse prio thread:
  delete before start: succeeded -> result: thread id
  delete after delete: failed -> result: -1
  delete after start: failed -> result: -1
  delete after suspend: failed -> result: -1
  delete after resume: failed -> result: -1
  delete after terminate: succeeded -> result: thread id
sleeping thread:
  delete after start: failed -> result: -1
  delete after suspend: failed -> result: -1
  delete after resume: failed -> result: -1
waiting thread:
  delete after start: failed -> result: -1
  delete after suspend: failed -> result: -1
  delete after resume: failed -> result: -1
corner cases:
  delete with invalid tid: failed -> result: -1
-- TEST END
